site/content/blog/identify-ghost-diff-during-kc-reconcile.md

```diff
@@ -0,0 +1,274 @@
+---
+title: "Identify ghost diff during kapp Controller reconciliation"
+slug: identify-ghost-diff-during-kapp-controller-reconciliation
+date: 2022-04-21
+author: Rohit Aggarwal
+excerpt: "Identify ghost diff during kapp Controller reconciliation "
+image: /img/logo.svg
+tags: ['carvel', 'kapp-controller', 'gitops', 'diffs', 'diff']
+---
+
+# Identify ghost diff during kapp Controller reconciliation
+
+[kapp controller](https://carvel.dev/kapp-controller/), a Package manager, is compatible with Gitops philosophy. It is continuously ensuring that the cluster is or converging towards the desired state. It does so by running the reconciliation loop after every `syncPeriod` duration. In each reconciliation cycle, it monitors the current state of the resources on the cluster and tries to bring it to the desired state if there is any mismatch. It does so with the help of [kapp](https://carvel.dev/kapp/). kapp, another carvel tool, tries to do a diff of the current live state of the resources with the desired state. Sometimes, even though there is no change in the desired state by the user explicitly, kapp still thinks some resources have changed and tries to redeploy them as part of the reconciliation. We call them `ghost` diffs. 
```

Comment:
Could we add bullet points for what can change in resources after they have been deployed.

Reply:
also, what happens due to ghost diffs? (impact, and why would we want to reduce them)

Reply:
Adopted. Added

Ghost diffs can appear due to different reasons: controller can change some configuration of the resource, HorizontalPodAutoscaler influencing the no. of replicas for a deployment, etc. Everytime a `ghost` diff is detected, a new configMap gets created. If you have many packages creating `ghost` diffs in the Kubernetes(K8s) cluster, you will end up having large no. of unnecessary configMaps which can slow the K8s API server response time.

Reply:
Adopted